oak/server/rust/oak_runtime/src/node/grpc/io.rs

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use log::{debug, error};
+use prost::Message;
+
+use oak_abi::{
+    proto::oak::encap::{GrpcRequest, GrpcResponse},
+    ChannelReadStatus, Handle, OakStatus,
+};
+
+use crate::{runtime::RuntimeProxy, NodeMessage};
+
+/// Wrapper for a [`Handle`] that is responsible for sending messages to an Oak channel.
+/// Uses [`RuntimeProxy`] to send messages to a specific Oak Node.
+pub struct Sender {
+    runtime: RuntimeProxy,
+    writer: Handle,
+}
+
+impl Sender {
+    pub fn new(runtime: RuntimeProxy, writer: Handle) -> Self {
+        Self { runtime, writer }
+    }
+
+    /// Composes a [`Message`] from `data` and `handles` and sends it to the [`Sender::writer`].
+    pub fn send(&self, data: Vec<u8>, handles: Vec<Handle>) -> Result<(), OakStatus> {
```

Comment:
The point of having the `NodeMessage` abstraction is to avoid having to pass around separate fields, so perhaps just use that directly? And implement `To` and `From` methods for things that need to be sent. Also it may make sense to just parametrize `Sender` and `Receiver` with a specific concrete type that can be converted to `NodeMessage`.

Basically I think this type should be the dual of the one in the SDK:

 https://github.com/project-oak/oak/blob/1f67e3b602f5ad2f72aa2be1c88412fcccedef33/sdk/rust/oak/src/io/receiver.rs#L32

In fact it may be worth a comment to clarify this too.

Reply:
Done